src/main/java/it/pagopa/pn/deliverypush/service/impl/LegalFactServiceImpl.java

```diff
@@ -96,42 +87,49 @@ private String readRecipientId( TimelineElementInternal timelineElement, Notific
-        }
-    }
+    public LegalFactDownloadMetadataResponse getLegalFactMetadata(String iun, LegalFactCategory legalFactType, String legalfactId) {
+        log.debug( "getLegalFactMetadata for iun={} and legalfactId={}", iun, legalfactId );
+        // la key è la legalfactid
+        FileDownloadResponse fileDownloadResponse = safeStorageClient.getFile(legalfactId, false);
+        LegalFactDownloadMetadataResponse response = new LegalFactDownloadMetadataResponse();
 
-    @NotNull
-    private ResponseEntity<Resource> getPaperFeedbackLegalFact(String iun, String legalfactId) {
-        final String attachmentId = legalfactId.replace("~","/");
-        log.debug( "Retrieve attachment url from External Channel with attachmentId={}", attachmentId );
-        String[] response = this.externalChannelClient.getResponseAttachmentUrl( new String[] {attachmentId} );
+        response.setFilename(buildLegalFactFilename(iun, legalFactType, legalfactId));
+        response.setContentLength(fileDownloadResponse.getContentLength());
+        response.setRetryAfter(fileDownloadResponse.getDownload().getRetryAfter());
+        response.setUrl(fileDownloadResponse.getDownload().getUrl());
 
-        if ( response != null && response.length > 0 ) {
-            try {
-                final UrlResource urlResource = new UrlResource(URI.create(response[0]));
-                return ResponseEntity.ok()
-                        .headers( fileStorage.headers() )
-                        .contentType(MediaType.APPLICATION_PDF)
-                        .body( urlResource );
+        return response;
+    }
 
-            } catch (MalformedURLException e) {
-                log.error( "Unable to retrieve a valid attachment url for iun={} and attachmentId={}",
-                        iun,
-                        attachmentId );
-                throw new PnInternalException( "Unable to retrieve resource " + response[0], e );
-            }
+    public ResponseEntity<Resource> getLegalfact(String iun, LegalFactCategory legalFactType, String legalfactId ) {
+        log.debug( "getLegalfact for iun={} and legalfactId={}", iun, legalfactId );
+        // la key è la legalfactid
+        FileDownloadResponse fileDownloadResponse = safeStorageClient.getFile(legalfactId, false);
 
-        } else {
-            final String message = String.format(MISSING_EXT_CHA_LEGAL_FACT_MESSAGE, iun, legalfactId);
-            log.error(message);
-            throw new PnInternalException(message);
+        try {
+            return ResponseEntity.ok()
+                    .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=" + buildLegalFactFilename(iun, legalFactType, legalfactId))
+                    .contentType(MediaType.parseMediaType(fileDownloadResponse.getContentType()))
+                    .body( new UrlResource(fileDownloadResponse.getDownload().getUrl()) );
+        } catch (MalformedURLException e) {
+            log.error("cannot stream content", e);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This code contains a potential information leakage in the handling of the following *logged data*: (MalformedURLException e). Logging an exception object typically results in printing the full stack trace, which reveals variable names, method names, line numbers, and other parts of the code.

**Fix**
Release data more selectively, or declassify the data before it&#39;s released. Alternatively, switch to the DEBUG (debug() API) or TRACE (trace() API) logging levels.

**More info**
[Learn more about information leakage](https://cwe.mitre.org/data/definitions/532.html) (external link).